pkg/socketservice/helm.go

```diff
@@ -0,0 +1,172 @@
+package socketservice
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"regexp"
+
+	"github.com/mholt/archiver"
+	"github.com/pkg/errors"
+)
+
+var (
+	goTemplateRegex *regexp.Regexp
+)
+
+func init() {
+	goTemplateRegex = regexp.MustCompile(`({{)|(}})`)
+}
+
+// When saving templated file back into a chart, we need to escape Go templates so second Helm pass would ignore them.
+// These are appication templates that maybe used in application config files and Helm should ignore them.
+// For example, origianl chart has this:
+//		"legendFormat": "{{`{{`}} value {{`}}`}}",
+// Rendered chart becomes:
+//		"legendFormat": "{{ value }}",
+// Repackaged chart should have this:
+//		"legendFormat": "{{`{{`}} value {{`}}`}}",
+func escapeGoTemplates(content []byte) []byte {
+	replace := func(in []byte) []byte {
+		if string(in) == "{{" {
+			return []byte("{{`{{`}}")
+		}
+		if string(in) == "}}" {
+			return []byte("{{`}}`}}")
+		}
+		return in
+	}
+
+	return goTemplateRegex.ReplaceAllFunc(content, replace)
+}
+
+func renderChartsArchive(deployedVersionArchive string, name string, version string) ([]byte, error) {
+	archiveChartDir := filepath.Join(deployedVersionArchive, "overlays", "downstreams", name, "charts")
+	_, err := os.Stat(archiveChartDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	exportChartPath, err := ioutil.TempDir("", "kotsadm")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create temp dir")
+	}
+	defer os.RemoveAll(exportChartPath)
+
+	desrChartsDir := filepath.Join(exportChartPath, "charts")
+	if _, err := os.Stat(desrChartsDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(desrChartsDir, 0744); err != nil {
+			return nil, errors.Wrap(err, "failed to mkdir for archive chart")
+		}
+	}
+
+	sourceChartsDir := filepath.Join(deployedVersionArchive, "base", "charts")
+	metadataFiles := []string{"Chart.yaml", "Chart.lock"}
```

Comment:
i wonder if we need any other files

https://github.com/helm/helm/blob/21251378b432fb629cb1a49af3892e20c1ec9624/pkg/chart/loader/load.go#L95-L155

not really for this pr but something to think about.

Reply:
interesting.  the requirements files might be of interest.  i don't know what's in those.  but values should be used on initial render and can be discarded.